userspace/libsinsp/container_engine/docker_common.cpp

```diff
@@ -814,7 +814,29 @@ bool docker_async_source::parse_docker(const docker_async_instruction& instructi
 	{
 		container.m_cpu_shares = cpu_shares;
 	}
-	container.m_cpu_quota = host_config_obj["CpuQuota"].asInt64();
+
+	/**
+	 * 2 separate docker APIs use CFS CPU scheduler to constrain container CPU usage
+	 * Reference: https://docs.docker.com/engine/reference/run/
+	 * 1) docker run --cpus=<Number of CPUS>
+	 *    <Number of CPUs> is converted into a cfs_cpu_quota value for the default cfs_cpu_period=100000
+	 *    cfs_cpu_period cannot be changed with this API
+	 *    For example, if <Number of CPUs>=0.5, cfs_cpu_quota=50000 and cfs_cpu_period=100000
+	 * 2) docker run --cpu-quota=<quota> --cpu-period=<period>
+	 *    CFS quota and/or period can be set directly. The default period is 100000 and default quota
+	 *    is 0 (which translates to unconstrained)
+	 *    For example, if <quota>=12345 and <period>=67890, then cfs_cpu_quota=12345 and cfs_cpu_period=67890
+	 * These 2 APIs are mutually exclusive: docker throws an error if an attempt is made to use --cpus in combination
+	 * with either --cpu-quota or --cpu-period
+	 *
+	 * docker_response json output:
+	 * 1) When --cpus is used, the value is returned as NanoCpus; both CpuQuota and CpuPeriod are 0
+	 *    Since cfs_cpu_period=100000=10^5 and 10^9 NanoCpus is 1 CPU, which translates to cfs_cpu_quota=100000=10^5,
+	 *    we need to divide NanoCpus by 10^4=10000 to convert NanoCpus into cfs_cpu_quota
+	 *
+	 * 2) When --cpu-quota and/or --cpu-period are used, the corresponding values are returned; NanoCpus is 0
+	 */
+	container.m_cpu_quota = max(host_config_obj["CpuQuota"].asInt64(), host_config_obj["NanoCpus"].asInt64()/10000);
```

Comment:
From reading your comment, this is the number of cpus in the cpuset. That should already be coming back in `host_config_obj["CpusetCpus"]`.

But if this is the right thing to do, then we must also set CpuPeriod because every consumer is likely to do CpuQuota/CpuPeriod to get the number of cpus.

Reply:
Naveen pointed out I am wrong about cpusets, that is `--cpuset-cpus` not `--cpus`

So I am ok with this change. I sorta think that you should explicitly set `m_cpu_period` for the NanoCpus because it just looks odd to set CpuQuota without setting CpuPeriod (and relying on the ctor value). Defer to developer there.

Reply:
m_cpu_period is initialized in the container_info constructor to the default value of 100000 
https://github.com/draios/sysdig/blob/6015e114905614db84a2676a668b31c17ef2d49f/userspace/libsinsp/container_info.h#L197

It then is updated immediately below my change (existing code) https://github.com/draios/sysdig/blob/6015e114905614db84a2676a668b31c17ef2d49f/userspace/libsinsp/container_engine/docker_common.cpp#L821

~~~~~~~
	const auto cpu_period = host_config_obj["CpuPeriod"].asInt64();
	if(cpu_period > 0)
	{
		container.m_cpu_period = cpu_period;
	}
~~~~~~~

The code is written this way because docker sends 0 as the period value in json when it is the default value.
We only want to pick up non-default values here and those are actually filled in by docker in the response. And, when --cpus is used there is no way to get a non-default period value (as I explained in the comment I added to the code).

Reply:
I left this as is because this behavior is not specific to the case when we have to pick up quota from NanoCpus. If docker run --cpu-quota is run without passing in a --cpu-period then CpuQuota is filled in but CpuPeriod is not.